File: crates/spectra-http/src/anthropic.rs
use async_trait::async_trait;
use futures_util::StreamExt;
use reqwest::Client;
use serde::Deserialize;
use tokio::sync::mpsc;
use tokio_stream::wrappers::ReceiverStream;
use spectra_rs::error::{Result, SpectraError};
use spectra_rs::event::ContentDelta;
use spectra_rs::llm::{LlmClient, LlmStream, LlmStreamEvent, LlmRequest, LlmResponse, Provider};
use spectra_rs::messages::{
    AssistantMessage, Content, Message, StopReason, TokenUsage, ToolCall,
};

const ANTHROPIC_API_URL: &str = "https://api.anthropic.com/v1/messages";
const DEFAULT_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(120);
const CONNECT_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(30);

pub struct AnthropicClient {
    client: Client,
    api_key: Option<String>,
    base_url: Option<String>,
}

impl AnthropicClient {
    pub fn new(api_key: Option<String>) -> Result<Self> {
        let client = Client::builder()
            .use_rustls_tls()
            .timeout(DEFAULT_TIMEOUT)
            .connect_timeout(CONNECT_TIMEOUT)
            .pool_max_idle_per_host(10)
            .build()
            .map_err(|e| SpectraError::LlmError {
                provider: "anthropic".into(),
                message: format!("Failed to create HTTP client: {}", e),
                source: Some(Box::new(e)),
            })?;
        Ok(Self { client, api_key, base_url: None })
    }

    pub fn with_api_key(api_key: impl Into<String>) -> Result<Self> {
        Self::new(Some(api_key.into()))
    }

    pub fn with_base_url(mut self, url: impl Into<String>) -> Self {
        self.base_url = Some(url.into());
        self
    }

    pub fn with_timeout(mut self, timeout: std::time::Duration) -> Result<Self> {
        self.client = Client::builder()
            .use_rustls_tls()
            .timeout(timeout)
            .connect_timeout(CONNECT_TIMEOUT)
            .pool_max_idle_per_host(10)
            .build()
            .map_err(|e| SpectraError::LlmError {
                provider: "anthropic".into(),
                message: format!("Failed to create HTTP client: {}", e),
                source: Some(Box::new(e)),
            })?;
        Ok(self)
    }

    fn get_api_key(&self) -> Result<String> {
        self.api_key
            .clone()
            .or_else(|| std::env::var("ANTHROPIC_API_KEY").ok())
            .ok_or_else(|| SpectraError::LlmError {
                provider: "anthropic".into(),
                message: "ANTHROPIC_API_KEY not set".into(),
                source: None,
            })
    }

    fn build_request_body(&self, request: &LlmRequest) -> Result<String> {
        let mut body = serde_json::Map::new();
        body.insert("model".into(), serde_json::Value::String(request.model.id.clone()));
        body.insert("max_tokens".into(), serde_json::Value::Number(request.model.config.max_tokens.into()));
        body.insert("stream".into(), serde_json::Value::Bool(true));

        if let Some(temp) = request.model.config.temperature {
            body.insert("temperature".into(), serde_json::json!(temp));
        }

        let mut messages: Vec<serde_json::Value> = request.messages.iter().map(|msg| {
            match msg {
                Message::User(u) => serde_json::json!({
                    "role": "user",
                    "content": user_content_to_json(&u.content),
                }),
                Message::Assistant(a) => serde_json::json!({
                    "role": "assistant",
                    "content": assistant_content_to_json(&a.content, &a.tool_calls),
                }),
                Message::ToolResult(tr) => serde_json::json!({
                    "role": "user",
                    "content": [{
                        "type": "tool_result",
                        "tool_use_id": tr.tool_call_id,
                        "content": tr.content.to_string(),
                    }],
                }),
            }
        }).collect();

        let cache_control = build_cache_control(&self.base_url);

        // Add cache_control to the last content block of the last user message
        if let Some(ref cc) = cache_control {
            if let Some(last_msg) = messages.last_mut() {
                if last_msg.get("role").and_then(|r| r.as_str()) == Some("user") {
                    if let Some(content) = last_msg.get_mut("content") {
                        add_cache_control_to_last_block(content, cc);
                    }
                }
            }
        }

        body.insert("messages".into(), serde_json::Value::Array(messages));

        if let Some(system) = &request.system_prompt {
            if let Some(ref cc) = cache_control {
                body.insert("system".into(), serde_json::json!([{
                    "type": "text",
                    "text": system,
                    "cache_control": cc
                }]));
            } else {
                body.insert("system".into(), serde_json::Value::String(system.clone()));
            }
        }

        if !request.tools.is_empty() {
            let tools: Vec<serde_json::Value> = request.tools.iter().map(|t| {
                serde_json::json!({
                    "name": t.name,
                    "description": t.description,
                    "input_schema": t.parameters,
                })
            }).collect();
            body.insert("tools".into(), serde_json::Value::Array(tools));
        }

        serde_json::to_string(&body).map_err(|e| e.into())
    }

    pub async fn stream_request(&self, request: LlmRequest) -> Result<LlmStream> {
        let api_key = self.get_api_key()?;
        let body = self.build_request_body(&request)?;
        let (tx, rx) = mpsc::channel(256);
        let client = self.client.clone();
        let url = self.base_url.clone().unwrap_or_else(|| ANTHROPIC_API_URL.to_string());

        let model_id = request.model.id.clone();
        let provider_name = "anthropic".to_string();

        tokio::spawn(async move {
            let mut assistant_msg = AssistantMessage::new(Vec::new(), Vec::new(), StopReason::EndOfTurn);
            assistant_msg.provider = provider_name;
            assistant_msg.model = model_id;
            let mut current_tool: Option<ToolCall> = None;
            let mut in_tool = false;

            let _ = tx.send(Ok(LlmStreamEvent::Start { partial: assistant_msg.clone() })).await;

            let response = match client
                .post(&url)
                .header("content-type", "application/json")
                .header("x-api-key", &api_key)
                .header("anthropic-version", "2023-06-01")
                .header("accept", "text/event-stream")
                .body(body)
                .send()
                .await
            {
                Ok(r) => r,
                Err(e) => {
                    let _ = tx.send(Err(SpectraError::LlmError {
                        provider: "anthropic".into(),
                        message: format!("Request failed: {}", e),
                        source: None,
                    })).await;
                    return;
                }
            };

            if !response.status().is_success() {
                let status = response.status();
                let body_text = response.text().await.unwrap_or_default();
                let _ = tx.send(Err(SpectraError::LlmError {
                    provider: "anthropic".into(),
                    message: format!("API error {}: {}", status, body_text),
                    source: None,
                })).await;
                return;
            }

            let mut stream = response.bytes_stream();
            let mut line = Vec::new();

            while let Some(chunk) = stream.next().await {
                match chunk {
                    Ok(bytes) => {
                        for byte in bytes {
                            match byte {
                                b'\n' | b'\r' if !line.is_empty() => {
                                    let s = String::from_utf8_lossy(&line).trim().to_string();
                                    line.clear();
                                    if !s.starts_with("data: ") || s.len() <= 6 { continue; }
                                    let data = &s[6..];
                                    if data == "[DONE]" { break; }
                                    parse_event(data, &mut assistant_msg, &mut current_tool, &mut in_tool, &tx).await;
                                }
                                _ => line.push(byte),
                            }
                        }
                    }
                    Err(e) => {
                        let _ = tx.send(Err(SpectraError::LlmError {
                            provider: "anthropic".into(),
                            message: format!("Stream error: {}", e),
                            source: None,
                        })).await;
                        break;
                    }
                }
            }

            for tc in &mut assistant_msg.tool_calls {
                if let serde_json::Value::String(s) = &tc.arguments {
                    if let Ok(parsed) = serde_json::from_str::<serde_json::Value>(s) {
                        tc.arguments = parsed;
                    }
                }
            }

            let _ = tx.send(Ok(LlmStreamEvent::Done { message: assistant_msg })).await;
        });

        Ok(Box::pin(ReceiverStream::new(rx)))
    }
}

fn user_content_to_json(content: &[Content]) -> serde_json::Value {
    let items: Vec<serde_json::Value> = content.iter().filter_map(|c| {
        match c {
            Content::Text { text } => {
                if text.is_empty() { None } else {
                    Some(serde_json::json!({ "type": "text", "text": text }))
                }
            }
            Content::Image { url, .. } => Some(serde_json::json!({
                "type": "image",
                "source": { "type": "url", "url": url },
            })),
            Content::Thinking { .. } => None,
        }
    }).collect();
    if items.len() == 1 { items.into_iter().next().unwrap() } else { serde_json::json!(items) }
}

fn assistant_content_to_json(content: &[Content], tool_calls: &[ToolCall]) -> serde_json::Value {
    let mut items: Vec<serde_json::Value> = Vec::with_capacity(content.len() + tool_calls.len());

    for c in content {
        if let Content::Text { text } = c
            && !text.is_empty() {
                items.push(serde_json::json!({ "type": "text", "text": text }));
            }
    }

    for tc in tool_calls {
        items.push(serde_json::json!({
            "type": "tool_use",
            "id": tc.id,
            "name": tc.name,
            "input": tc.arguments,
        }));
    }

    if items.len() == 1 { items.into_iter().next().unwrap() } else { serde_json::json!(items) }
}

fn build_cache_control(base_url: &Option<String>) -> Option<serde_json::Value> {
    let retention = std::env::var("SPECTRA_CACHE_RETENTION").unwrap_or_else(|_| "short".into());
    if retention == "none" {
        return None;
    }
    let is_direct = base_url.as_ref().map_or(true, |u| u.contains("api.anthropic.com"));
    if retention == "long" && is_direct {
        Some(serde_json::json!({ "type": "ephemeral", "ttl": "1h" }))
    } else {
        Some(serde_json::json!({ "type": "ephemeral" }))
    }
}

fn add_cache_control_to_last_block(content: &mut serde_json::Value, cache_control: &serde_json::Value) {
    match content {
        serde_json::Value::Array(blocks) => {
            if let Some(last) = blocks.last_mut() {
                if let Some(obj) = last.as_object_mut() {
                    if obj.get("type").and_then(|t| t.as_str()) == Some("text")
                        || obj.get("type").and_then(|t| t.as_str()) == Some("tool_result")
                    {
                        obj.insert("cache_control".into(), cache_control.clone());
                    }
                }
            }
        }
        serde_json::Value::String(text) => {
            *content = serde_json::json!([{
                "type": "text",
                "text": text,
                "cache_control": cache_control
            }]);
        }
        _ => {}
    }
}

fn parse_stop_reason(reason: &str) -> StopReason {
    match reason {
        "end_turn" => StopReason::EndOfTurn,
        "max_tokens" => StopReason::MaxTokens,
        "tool_use" => StopReason::ToolCalls,
        _ => StopReason::EndOfTurn,
    }
}

#[derive(Deserialize)]
#[allow(dead_code)]
struct SSEEvent {
    #[serde(rename = "type")]
    event_type: String,
    #[serde(default)]
    index: Option<u32>,
    #[serde(default)]
    delta: Option<serde_json::Value>,
    #[serde(default)]
    content_block: Option<serde_json::Value>,
    #[serde(default)]
    message: Option<serde_json::Value>,
}

async fn parse_event(
    data: &str,
    msg: &mut AssistantMessage,
    current_tool: &mut Option<ToolCall>,
    in_tool: &mut bool,
    tx: &mpsc::Sender<std::result::Result<LlmStreamEvent, SpectraError>>,
) {
    if let Ok(event) = serde_json::from_str::<SSEEvent>(data) {
        match event.event_type.as_str() {
            "message_start" => {
                if let Some(msg_data) = &event.message {
                    if let Some(id) = msg_data.get("id").and_then(|v| v.as_str()) {
                        msg.response_id = Some(id.to_string());
                    }
                }
            }
            "content_block_start" => {
                if let Some(block) = event.content_block {
                    match block.get("type").and_then(|t| t.as_str()) {
                        Some("tool_use") => {
                            *in_tool = true;
                            let id = block.get("id").and_then(|v| v.as_str()).unwrap_or("").to_string();
                            let name = block.get("name").and_then(|v| v.as_str()).unwrap_or("").to_string();
                            *current_tool = Some(ToolCall {
                                id: id.clone(),
                                name: name.clone(),
                                arguments: serde_json::Value::Null,
                                thinking_signature: None,
                            });
                            let _ = tx.send(Ok(LlmStreamEvent::ContentDelta {
                                delta: ContentDelta::ToolCallStart { id, name },
                            })).await;
                        }
                        Some("thinking") => {
                            let thinking = block.get("thinking").and_then(|v| v.as_str()).unwrap_or("").to_string();
                            let signature = block.get("signature").and_then(|v| v.as_str()).map(|s| s.to_string());
                            msg.content.push(Content::Thinking {
                                thinking: thinking.clone(),
                                signature: signature.clone(),
                                redacted: false,
                            });
                            let _ = tx.send(Ok(LlmStreamEvent::ContentDelta {
                                delta: ContentDelta::Thinking { delta: thinking, signature },
                            })).await;
                        }
                        _ => {}
                    }
                }
            }
            "content_block_delta" => {
                if let Some(delta) = event.delta {
                    if *in_tool {
                        if let Some(text) = delta.get("partial_json").and_then(|t| t.as_str()) {
                            if let Some(tc) = current_tool.as_mut() {
                                if let serde_json::Value::String(s) = &mut tc.arguments {
                                    s.push_str(text);
                                } else {
                                    tc.arguments = serde_json::Value::String(text.to_string());
                                }
                                let _ = tx.send(Ok(LlmStreamEvent::ContentDelta {
                                    delta: ContentDelta::ToolCallDelta {
                                        id: tc.id.clone(),
                                        args_delta: text.to_string(),
                                    },
                                })).await;
                            }
                        }
                    } else if let Some(text) = delta.get("text").and_then(|t| t.as_str()) {
                        msg.content.push(Content::Text { text: text.to_string() });
                        let _ = tx.send(Ok(LlmStreamEvent::ContentDelta {
                            delta: ContentDelta::Text { delta: text.to_string() },
                        })).await;
                    } else if let Some(thinking_delta) = delta.get("thinking").and_then(|t| t.as_str()) {
                        let sig = delta.get("signature").and_then(|v| v.as_str()).map(|s| s.to_string());
                        msg.content.push(Content::Thinking {
                            thinking: thinking_delta.to_string(),
                            signature: sig.clone(),
                            redacted: false,
                        });
                        let _ = tx.send(Ok(LlmStreamEvent::ContentDelta {
                            delta: ContentDelta::Thinking { delta: thinking_delta.to_string(), signature: sig },
                        })).await;
                    }
                }
            }
            "content_block_stop" => {
                if *in_tool {
                    if let Some(tc) = current_tool.take() {
                        msg.tool_calls.push(tc.clone());
                        let _ = tx.send(Ok(LlmStreamEvent::ContentDelta {
                            delta: ContentDelta::ToolCallEnd { id: tc.id.clone() },
                        })).await;
                    }
                    *in_tool = false;
                }
            }
            "message_delta" => {
                if let Some(delta) = event.delta {
                    if let Some(reason) = delta.get("stop_reason").and_then(|r| r.as_str()) {
                        msg.stop_reason = parse_stop_reason(reason);
                    }
                    if let Some(usage) = delta.get("usage") {
                        msg.usage.input_tokens = usage.get("input_tokens").and_then(|v| v.as_u64()).unwrap_or(0) as u32;
                        msg.usage.output_tokens = usage.get("output_tokens").and_then(|v| v.as_u64()).unwrap_or(0) as u32;
                        msg.usage.cache_read_tokens = usage.get("cache_read_input_tokens").and_then(|v| v.as_u64()).unwrap_or(0) as u32;
                        msg.usage.cache_write_tokens = usage.get("cache_creation_input_tokens").and_then(|v| v.as_u64()).unwrap_or(0) as u32;
                    }
                }
            }
            "message_stop" => {
                for tc in &mut msg.tool_calls {
                    if let serde_json::Value::String(s) = &tc.arguments {
                        if let Ok(parsed) = serde_json::from_str::<serde_json::Value>(s) {
                            tc.arguments = parsed;
                        }
                    }
                }
            }
            _ => {}
        }
    }
}

#[async_trait]
impl LlmClient for AnthropicClient {
    async fn complete(&self, request: LlmRequest) -> Result<LlmResponse> {
        let mut stream = self.stream(request).await?;
        while let Some(event) = stream.next().await {
            match event? {
                LlmStreamEvent::Done { message } => {
                    let stop = message.stop_reason;
                    return Ok(LlmResponse { message, usage: TokenUsage::default(), stop_reason: stop });
                }
                LlmStreamEvent::Error { message } => {
                    return Err(SpectraError::LlmError { provider: "anthropic".into(), message, source: None });
                }
                _ => continue,
            }
        }
        Err(SpectraError::LlmError {
            provider: "anthropic".into(),
            message: "No response from Anthropic".into(),
            source: None,
        })
    }

    async fn stream(&self, request: LlmRequest) -> Result<LlmStream> {
        self.stream_request(request).await
    }

    fn provider(&self) -> Provider {
        Provider::Anthropic
    }
}

File: packages/ai/src/providers/anthropic.ts
import Anthropic from '@anthropic-ai/sdk';
import type {
	MessageParam,
	MessageCreateParamsStreaming,
	TextBlockParam,
	ImageBlockParam,
	ToolUseBlockParam,
	ToolResultBlockParam,
} from '@anthropic-ai/sdk/resources/messages.js';
import type {
	AssistantMessage,
	Context,
	Message,
	Model,
	StreamOptions,
	TextContent,
	ThinkingContent,
	Tool,
	ToolCall,
} from '../types.js';
import { AssistantMessageEventStream } from '../event-stream.js';
import { sanitizeSurrogates } from './shared.js';

type ContentBlockParam = TextBlockParam | ImageBlockParam | ToolUseBlockParam | ToolResultBlockParam;
type StreamBlock = (ThinkingContent | TextContent | ToolCall) & { index?: number; partialJson?: string };

const THINKING_BUDGETS: Record<string, number> = {
	low: 2048,
	medium: 8192,
	high: 16000,
	max: 31999,
};

function getEnvApiKey(provider: string): string | undefined {
	const keys: Record<string, string | undefined> = {
		anthropic: process.env.ANTHROPIC_API_KEY ?? process.env.ANTHROPIC_KEY,
	};
	return keys[provider];
}

function toAnthropicMessage(message: Message): MessageParam {
	if (message.role === 'user') {
		const content =
			typeof message.content === 'string'
				? sanitizeSurrogates(message.content)
				: message.content.map((block) => {
						if (block.type === 'text') {
							return { type: 'text' as const, text: sanitizeSurrogates(block.text) };
						}
						if (block.type === 'image') {
							return {
								type: 'image' as const,
								source: {
									type: 'base64' as const,
									media_type: block.mimeType as 'image/jpeg' | 'image/png' | 'image/gif' | 'image/webp',
									data: block.data,
								},
							};
						}
						return { type: 'text' as const, text: '' };
					});
		return { role: 'user', content };
	}

	if (message.role === 'assistant') {
		const content: ContentBlockParam[] = message.content.map((block) => {
			if (block.type === 'text') {
				return { type: 'text' as const, text: sanitizeSurrogates(block.text) };
			}
			if (block.type === 'thinking') {
				return {
					type: 'text' as const,
					text: sanitizeSurrogates(block.thinking),
				};
			}
			if (block.type === 'toolCall') {
				let input: Record<string, unknown>;
				if (typeof block.arguments === 'string') {
					try {
						input = JSON.parse(block.arguments);
					} catch {
						input = {};
					}
				} else {
					input = block.arguments;
				}
				return {
					type: 'tool_use' as const,
					id: block.id,
					name: block.name,
					input,
				};
			}
			return { type: 'text' as const, text: '' };
		});
		return { role: 'assistant', content };
	}

	if (message.role === 'toolResult') {
		const content: ContentBlockParam[] = message.content.map((block) => {
			if (block.type === 'text') {
				return {
					type: 'tool_result' as const,
					tool_use_id: message.toolCallId,
					content: sanitizeSurrogates(block.text),
				};
			}
			if (block.type === 'image') {
				return {
					type: 'tool_result' as const,
					tool_use_id: message.toolCallId,
					content: [
						{
							type: 'image' as const,
							source: {
								type: 'base64' as const,
								media_type: block.mimeType as 'image/jpeg' | 'image/png' | 'image/gif' | 'image/webp',
								data: block.data,
							},
						},
					],
				};
			}
			return { type: 'tool_result' as const, tool_use_id: message.toolCallId, content: '' };
		});
		return { role: 'user', content };
	}

	return { role: 'user', content: '' };
}

function toAnthropicTool(tool: Tool): {
	name: string;
	description: string;
	input_schema: { type: 'object'; properties: Record<string, unknown>; required: string[] };
} {
	const jsonSchema = tool.parameters as { properties?: Record<string, unknown>; required?: string[] };
	return {
		name: tool.name,
		description: tool.description,
		input_schema: {
			type: 'object',
			properties: jsonSchema.properties ?? {},
			required: jsonSchema.required ?? [],
		},
	};
}

type CacheRetention = 'short' | 'long' | 'none';

function getCacheControl(baseUrl?: string): { type: 'ephemeral'; ttl?: '1h' } | undefined {
	const retention = (process.env.SPECTRA_CACHE_RETENTION as CacheRetention) || 'short';
	if (retention === 'none') return undefined;
	const ttl = retention === 'long' && (!baseUrl || baseUrl.includes('api.anthropic.com')) ? '1h' : undefined;
	return { type: 'ephemeral', ...(ttl && { ttl }) };
}

function applyCacheControl(messages: MessageParam[], cacheControl?: { type: 'ephemeral'; ttl?: '1h' }): void {
	if (!cacheControl) return;
	for (let i = messages.length - 1; i >= 0; i--) {
		const msg = messages[i];
		if (msg.role !== 'user') continue;

		const content = msg.content;
		if (typeof content === 'string') {
			msg.content = [{ type: 'text' as const, text: content, cache_control: cacheControl }] as any;
			return;
		}

		if (Array.isArray(content) && content.length > 0) {
			const last = content[content.length - 1] as any;
			if (last && (last.type === 'text' || last.type === 'tool_result')) {
				last.cache_control = cacheControl;
			}
			return;
		}
	}
}

function mapStopReason(reason: string): 'stop' | 'length' | 'toolUse' | 'error' | 'aborted' {
	switch (reason) {
		case 'end_turn':
			return 'stop';
		case 'max_tokens':
			return 'length';
		case 'tool_use':
			return 'toolUse';
		case 'refusal':
		case 'sensitive':
			return 'error';
		default:
			return 'stop';
	}
}

export function createAnthropicProvider() {
	return {
		name: 'anthropic',
		listModels: () => import('../models.js').then((m) => m.getProviderModels('anthropic')),
		stream(model: Model, context: Context, options?: StreamOptions): AssistantMessageEventStream {
			const stream = new AssistantMessageEventStream();

			const run = async () => {
				const apiKey = options?.apiKey ?? getEnvApiKey(model.provider);

				if (!apiKey) {
					const errorMsg: AssistantMessage = {
						role: 'assistant',
						content: [],
						provider: model.provider,
						model: model.id,
						responseId: undefined,
						usage: { input: 0, output: 0, cacheRead: 0, cacheWrite: 0, totalTokens: 0 },
						stopReason: 'error',
						errorMessage: 'ANTHROPIC_API_KEY not set',
						timestamp: Date.now(),
					};
					stream.push({ type: 'start', partial: errorMsg });
					stream.push({ type: 'error', reason: 'error', error: errorMsg });
					stream.end();
					return;
				}

				const client = new Anthropic({ apiKey });

				const output: AssistantMessage = {
					role: 'assistant',
					content: [],
					provider: model.provider,
					model: model.id,
					responseId: undefined,
					usage: { input: 0, output: 0, cacheRead: 0, cacheWrite: 0, totalTokens: 0 },
					stopReason: 'stop',
					timestamp: Date.now(),
				};

				try {
					const messages: MessageParam[] = context.messages.map(toAnthropicMessage);
					const tools = context.tools?.map(toAnthropicTool) ?? [];

					const thinkingEffort = options?.thinkingEffort;
					const budget =
						thinkingEffort && thinkingEffort !== 'none' ? THINKING_BUDGETS[thinkingEffort] : undefined;

				const cacheControl = getCacheControl(model.baseUrl);

				const systemBlocks = context.systemPrompt
					? [{ type: 'text' as const, text: sanitizeSurrogates(context.systemPrompt), ...(cacheControl && { cache_control: cacheControl }) }]
					: undefined;

				applyCacheControl(messages, cacheControl);

				const params = {
					model: model.id,
					max_tokens: budget ? (options?.maxTokens ?? 8192) : (options?.maxTokens ?? 4096),
					messages,
					tools: tools.length > 0 ? tools : undefined,
					system: systemBlocks,
					temperature: options?.temperature,
					thinking: budget ? { type: 'enabled', budget_tokens: budget } : undefined,
					stream: true,
				} as MessageCreateParamsStreaming;

					const anthropicStream = client.messages.stream(params, { signal: options?.signal });
					stream.push({ type: 'start', partial: output });

					const blocks = output.content as StreamBlock[];

					for await (const event of anthropicStream) {
						if (event.type === 'message_start') {
							output.responseId = event.message.id;
							output.usage.input = event.message.usage.input_tokens || 0;
							output.usage.output = event.message.usage.output_tokens || 0;
							output.usage.cacheRead =
								(event.message.usage as { cache_read_input_tokens?: number }).cache_read_input_tokens || 0;
							output.usage.cacheWrite =
								(event.message.usage as { cache_creation_input_tokens?: number }).cache_creation_input_tokens ||
								0;
							output.usage.totalTokens =
								output.usage.input + output.usage.output;
						} else if (event.type === 'content_block_start') {
							if (event.content_block.type === 'text') {
								const block: StreamBlock = { type: 'text', text: '', index: event.index };
								output.content.push(block);
								stream.push({ type: 'text_start', contentIndex: output.content.length - 1, partial: output });
							} else if ((event.content_block as any).type === 'thinking') {
								const thinkingText = (event.content_block as any).thinking || '';
								const block: StreamBlock = { type: 'thinking', thinking: thinkingText, index: event.index };
								output.content.push(block);
								stream.push({
									type: 'thinking_start',
									contentIndex: output.content.length - 1,
									partial: output,
								});
							} else if ((event.content_block as any).type === 'redacted_thinking') {
								const block: StreamBlock = {
									type: 'thinking',
									thinking: '[Redacted thinking]',
									redacted: true,
									index: event.index,
								};
								output.content.push(block);
								stream.push({
									type: 'thinking_start',
									contentIndex: output.content.length - 1,
									partial: output,
								});
							} else if (event.content_block.type === 'tool_use') {
								const block: StreamBlock = {
									type: 'toolCall',
									id: event.content_block.id,
									name: event.content_block.name,
									arguments: (event.content_block.input as Record<string, unknown>) ?? {},
									partialJson: '',
									index: event.index,
								};
								output.content.push(block);
								stream.push({
									type: 'toolcall_start',
									contentIndex: output.content.length - 1,
									partial: output,
								});
							}
						} else if (event.type === 'content_block_delta') {
							if (event.delta.type === 'text_delta') {
								const idx = blocks.findIndex((b) => b.index === event.index);
								const block = blocks[idx];
								if (block && block.type === 'text') {
									block.text += event.delta.text;
									stream.push({
										type: 'text_delta',
										contentIndex: idx,
										delta: event.delta.text,
										partial: output,
									});
								}
							} else if ((event.delta as any).type === 'thinking_delta') {
								const idx = blocks.findIndex((b) => b.index === event.index);
								const block = blocks[idx];
								if (block && block.type === 'thinking') {
									block.thinking += (event.delta as any).thinking || '';
									stream.push({
										type: 'thinking_delta',
										contentIndex: idx,
										delta: (event.delta as any).thinking || '',
										partial: output,
									});
								}
							} else if (event.delta.type === 'input_json_delta') {
								const idx = blocks.findIndex((b) => b.index === event.index);
								const block = blocks[idx];
								if (block && block.type === 'toolCall') {
									block.partialJson = (block.partialJson ?? '') + event.delta.partial_json;
									try {
										block.arguments = JSON.parse(block.partialJson);
									} catch {
										// incomplete JSON, keep partial
									}
									stream.push({
										type: 'toolcall_delta',
										contentIndex: idx,
										delta: event.delta.partial_json,
										partial: output,
									});
								}
							}
						} else if (event.type === 'content_block_stop') {
							const idx = blocks.findIndex((b) => b.index === event.index);
							const block = blocks[idx];
							if (block) {
								delete block.index;
								if (block.type === 'text') {
									stream.push({ type: 'text_end', contentIndex: idx, content: block.text, partial: output });
								} else if (block.type === 'thinking') {
									stream.push({
										type: 'thinking_end',
										contentIndex: idx,
										content: block.thinking,
										partial: output,
									});
								} else if (block.type === 'toolCall') {
									try {
										block.arguments = JSON.parse(block.partialJson ?? '');
									} catch {
										block.arguments = {};
									}
									delete block.partialJson;
									stream.push({
										type: 'toolcall_end',
										contentIndex: idx,
										toolCall: block as ToolCall,
										partial: output,
									});
								}
							}
						} else if (event.type === 'message_delta') {
							if (event.delta.stop_reason) {
								output.stopReason = mapStopReason(event.delta.stop_reason);
							}
						}
					}

					if (options?.signal?.aborted) {
						throw new Error('Request aborted');
					}

					cleanupBlocks(blocks);

					stream.push({ type: 'done', reason: output.stopReason, message: output });
					stream.end();
				} catch (error) {
					cleanupOutputContent(output);
					output.stopReason = options?.signal?.aborted ? 'aborted' : 'error';
					output.errorMessage = error instanceof Error ? error.message : String(error);
					stream.push({ type: 'error', reason: output.stopReason, error: output });
					stream.end();
				}
			};

			run();

			return stream;
		},
	};
}

function cleanupBlocks(blocks: StreamBlock[]): void {
	for (const block of blocks) {
		delete block.index;
		delete block.partialJson;
	}
}

function cleanupOutputContent(output: AssistantMessage): void {
	for (const block of output.content) {
		delete (block as StreamBlock).index;
		delete (block as StreamBlock).partialJson;
	}
}
